app/src/main/scala/org/alephium/explorer/persistence/dao/BlockDao.scala

```diff
@@ -130,6 +133,47 @@ object BlockDao {
       .expireAfterWrite(5, java.util.concurrent.TimeUnit.SECONDS)
       .buildAsync(blockTimeAsyncLoader)
 
+    /** Caches the number of rows return by a query.
+      *
+      * @note Slick's [[Query]] type does not implement equals or hashCode so
+      *       this cache will not work as expected on dynamically generated queries.
+      *       The queries to cache should be static i.e. should have the same memory
+      *       address.
+      *
+      *        {{{
+      *          //OK
+      *          cacheRowCount.get(mainChainQuery)
+      *          cacheRowCount.get(mainChainQuery)
+      *
+      *          //OK
+      *          cacheRowCount.get(BlockHeaderSchema.table)
+      *          cacheRowCount.get(BlockHeaderSchema.table)
+      *
+      *          //NOT OK
+      *          cacheRowCount.get(BlockHeaderSchema.table.filter(_.mainChain))
+      *          cacheRowCount.get(BlockHeaderSchema.table.filter(_.mainChain))
+      *        }}}
+      * */
+    private val cacheRowCount: CaffeineAsyncCache[Query[_, _, Seq], Int] =
+      CaffeineAsyncCache {
+        Caffeine
+          .newBuilder()
+          .refreshAfterWrite(2, java.util.concurrent.TimeUnit.MINUTES)
+          .buildAsync[Query[_, _, Seq], Int] {
+            new AsyncCacheLoader[Query[_, _, Seq], Int] {
+              override def asyncLoad(table: Query[_, _, Seq],
+                                     executor: Executor): CompletableFuture[Int] =
+                run(table.length.result).asJava.toCompletableFuture
```

Comment:
I was first surprise to see the query to be the key of the cache, but in a sense it makes more sense that what I did to have a "fake" key : https://github.com/alephium/explorer-backend/blob/48986dd9b346aa26a05eb08f3016e14858a8e742/app/src/main/scala/org/alephium/explorer/persistence/dao/TransactionDao.scala#L55-L58 Can you maybe abstract what you did there so we could reuse it in the `TransactionDao` too?

Reply:
Yep :) [issue](https://github.com/alephium/explorer-backend/pull/180#issuecomment-1082603352) created